src/main/java/org/reactome/addlinks/referencecreators/KEGGReferenceCreator.java

```diff
@@ -95,48 +95,51 @@ public void createIdentifiers(long personID, Map<String, List<Map<KEGGKeys, Stri
-					targetDB = parts[0];
-					// now, the identifier should have been cleaned up by determineKeggReferenceDatabase
-					keggGeneIdentifier = parts[1];
-					xrefAlreadyExists = this.checkXRefExists(sourceReference, keggGeneIdentifier, targetDB);
-					if (!xrefAlreadyExists)
-					{
-						String keggDefinition = keggData.get(KEGGKeys.KEGG_DEFINITION);
-						sourceIdentifiersWithNewIdentifier++;
-						// Also need to add the keggDefinition and keggGeneID as "name" attributes.
-						List<String> names = new ArrayList<String>(3);
-						names.add(keggDefinition);
-						names.add(keggGeneIdentifier);
-						keggIdentifier = KEGGSpeciesCache.pruneKEGGSpeciesCode(keggIdentifier);
-						if (! keggIdentifier.equals(keggGeneIdentifier))
+//						// If the original KEGG_IDENTIFIER key didn't have a value, use the KEGG GENE ID.
+						if (keggIdentifier == null || keggIdentifier.trim().equals("") )
 						{
-							names.add(keggIdentifier);
+							keggIdentifier = keggGeneIdentifier;
 						}
-						Map<String,List<String>> extraAttributes = new HashMap<String,List<String>>(1);
-						extraAttributes.put(ReactomeJavaConstants.name, names);
-						logger.trace("For {}, creating new KEGG xref: {}",sourceReference.getDisplayName(), keggIdentifier);
-						if (!this.testMode)
+
+						logger.trace("Working on source object: {}", sourceReference.toString());
+						String targetDB = null;
+						KEGGReferenceCreatorHelper referenceCreatorHelper = new KEGGReferenceCreatorHelper(objectCache, logger);
+						String[] parts = referenceCreatorHelper.determineKeggReferenceDatabase(keggGeneIdentifier, keggPrefix);
+						targetDB = parts[0];
+						// now, the identifier should have been cleaned up by determineKeggReferenceDatabase
+						keggGeneIdentifier = parts[1];
+						xrefAlreadyExists = this.checkXRefExists(sourceReference, keggGeneIdentifier, targetDB);
+						if (!xrefAlreadyExists)
 						{
-							if (!this.testMode && targetDB != null)
+							String keggDefinition = keggData.get(KEGGKeys.KEGG_DEFINITION);
+							sourceIdentifiersWithNewIdentifier++;
+							// Also need to add the keggDefinition and keggGeneID as "name" attributes.
+							List<String> names = new ArrayList<>(3);
+							names.add(keggDefinition);
+							names.add(keggGeneIdentifier);
+							keggIdentifier = KEGGSpeciesCache.pruneKEGGSpeciesCode(keggIdentifier);
+							if (! keggIdentifier.equals(keggGeneIdentifier))
+							{
+								names.add(keggIdentifier);
+							}
+							Map<String,List<String>> extraAttributes = new HashMap<>(1);
+							extraAttributes.put(ReactomeJavaConstants.name, names);
+							logger.trace("For {}, creating new KEGG xref: {}",sourceReference.getDisplayName(), keggIdentifier);
+							if (!this.testMode)
 							{
-								this.refCreator.createIdentifier(keggGeneIdentifier, String.valueOf(sourceReference.getDBID()),targetDB, personID, this.getClass().getName(), speciesID, extraAttributes);
+								if (!this.testMode && targetDB != null)
```

Comment:
Unnecessary condition (`!this.testMode`) since it is checked right above

Reply:
Fixed.